tests/NTest/NTest_test.lua

```diff
@@ -0,0 +1,572 @@
+local N = require('NTest')("selftest")
```

Comment:
I'd almost prefer this file to be named `tests/test-ntest.lua` or something, so that `tests/NTest` was the library itself.... but maybe I'm being silly.

However, the use of `require` here along with the admonition to run NTest out of LFS would seem to mean that we also have to tell people to have `lua_examples/lfs/_init.lua` or equivalent on their devices under test.  Yes?

Reply:
With the file naming you got exactly the spot I was unsure about.
I think that in the tests folder there will be other tests over time with other test systems, namely your #2983.
Currently it seems as if the tests can be differentiated by their filetype but what if that does not hold true anymore.
I would only expect interaction between test systems when e.g. expect calls NTest.
So I would prefer having the different test systems in different folders.

How about 
/tests/NTest/system/NTest.lua
/tests/NTest/test_file.lua   // no - as this is no lua identifier and does not permit node.LUA.whatever() to work
or may be also file_test.lua tmr_test.lua

if they are all in tests and just named test_something it is unclear to which test system they belong.

Reply:
Does it matter which harness (if any) they're using?  I imagine the intended flow involves pushing *all* the requisite dependencies to the device under test... though I am sensitive to the possibility of host/DUT confusion if all the files are mixed together (as they presently are in #2983... I will fix that).  How about:

* `tests/host/expectnmcu` -- TCL support libraries from #2983
* `tests/host/utils` -- additional on-host test utilities, if anything emerges as such
* `tests/host/*.*` -- files that run on the host (in whatever language, but probably shell, Lua, and/or TCL)
* `tests/dut/NTest` -- this NTest framework
* `tests/dut/utils` -- additional on-DUT test libraries or utilities, if anything ever emerges in that category
* `tests/dut/*.lua` -- things to run on the DUT, named by the module they're testing, regardless of framework.

The broader test program, then, would be to make an LFS image containing at least all of `tests/dut/**` (recursively) and `lua_examples/lfs/_init.lua` (and maybe some other in-tree utilities, as needed) and lob that over onto the device (and blank its SPIFFS) before running tests, which then get run using `dofile()` as normal.

Reply:
But what if a test system can be run on both, the host and sut. NTest can alreldy run its selftest on the host and as soon as #3230 (or its reimplementation after Terry is done) goes in there will be more.

Btw: replaced the hint about big code and LFS with a copy from ftpserver which contains a link to LFS but also does not mention the need of _init.lua

Also changed the layout again. See checkin.

Reply:
That's fair and is a point I'd not considered.  I think I like your (implicit) proposal in the current commits to use `test/NTest/*` for NTest itself and `test/NTest_*.lua` for tests using NTest.  I'll rename my expect scripts to use a similar theme. :)